### ` Collaborative Filters`

In [1]:
import pandas as pd
import numpy as np 

In [2]:
u_col = ['user_id','age','sex','occupation','zip_code']

In [3]:
user = pd.read_csv('/Users/nandrajog/Downloads/ml-100k/u.user',sep ='|',names=u_col)

In [4]:
user

user_id  age sex     occupation zip_code
0          1   24   M     technician    85711
1          2   53   F          other    94043
2          3   23   M         writer    32067
3          4   24   M     technician    43537
4          5   33   F          other    15213
..       ...  ...  ..            ...      ...
938      939   26   F        student    33319
939      940   32   M  administrator    02215
940      941   20   M        student    97229
941      942   48   F      librarian    78209
942      943   22   M        student    77841

[943 rows x 5 columns]

In [5]:
i_cols = ['movie_id', 'title' ,'release date','video release date', 'IMDb URL', 'unknown', 'Action', 'Adventure',
 'Animation', 'Children\'s', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy',
 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller', 'War', 'Western']


movies = pd.read_csv('/Users/nandrajog/Downloads/ml-100k/u.item',sep='|', names = i_cols,encoding='latin-1')

In [6]:
movies

movie_id                                      title release date  \
0            1                           Toy Story (1995)  01-Jan-1995   
1            2                           GoldenEye (1995)  01-Jan-1995   
2            3                          Four Rooms (1995)  01-Jan-1995   
3            4                          Get Shorty (1995)  01-Jan-1995   
4            5                             Copycat (1995)  01-Jan-1995   
...        ...                                        ...          ...   
1677      1678                          Mat' i syn (1997)  06-Feb-1998   
1678      1679                           B. Monkey (1998)  06-Feb-1998   
1679      1680                       Sliding Doors (1998)  01-Jan-1998   
1680      1681                        You So Crazy (1994)  01-Jan-1994   
1681      1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996   

      video release date                                           IMDb URL  \
0                    NaN  http://us.imdb.com/M/title-exact?Toy%20Story%2...   
1                    NaN  http://us.imdb.com/M/title-exact?GoldenEye%20(...   
2                    NaN  http://us.imdb.com/M/title-exact?Four%20Rooms%...   
3                    NaN  http://us.imdb.com/M/title-exact?Get%20Shorty%...   
4                    NaN  http://us.imdb.com/M/title-exact?Copycat%20(1995)   
...                  ...                                                ...   
1677                 NaN  http://us.imdb.com/M/title-exact?Mat%27+i+syn+...   
1678                 NaN  http://us.imdb.com/M/title-exact?B%2E+Monkey+(...   
1679                 NaN      http://us.imdb.com/Title?Sliding+Doors+(1998)   
1680                 NaN  http://us.imdb.com/M/title-exact?You%20So%20Cr...   
1681                 NaN  http://us.imdb.com/M/title-exact?Schrei%20aus%...   

      unknown  Action  Adventure  Animation  Children's  ...  Fantasy  \
0           0       0          0          1           1  ...        0   
1           0       1          1          0           0  ...        0   
2           0       0          0          0           0  ...        0   
3           0       1          0          0           0  ...        0   
4           0       0          0          0           0  ...        0   
...       ...     ...        ...        ...         ...  ...      ...   
1677        0       0          0          0           0  ...        0   
1678        0       0          0          0           0  ...        0   
1679        0       0          0          0           0  ...        0   
1680        0       0          0          0           0  ...        0   
1681        0       0          0          0           0  ...        0   

      Film-Noir  Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  War  \
0             0       0        0        0        0       0         0    0   
1             0       0        0        0        0       0         1    0   
2             0       0        0        0        0       0         1    0   
3             0       0        0        0        0       0         0    0   
4             0       0        0        0        0       0         1    0   
...         ...     ...      ...      ...      ...     ...       ...  ...   
1677          0       0        0        0        0       0         0    0   
1678          0       0        0        0        1       0         1    0   
1679          0       0        0        0        1       0         0    0   
1680          0       0        0        0        0       0         0    0   
1681          0       0        0        0        0       0         0    0   

      Western  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
1677        0  
1678        0  
1679        0  
1680        0  
1681        0  

[1682 rows x 24 columns]

`“We see that this file gives us information regarding the movie's title, release date, IMDb URL, and its genre(s). Since we are focused on building only collaborative filters in this chapter, we do not require any of this information, apart from the movie title and its corresponding ID:”`



In [7]:
movies = movies[['movie_id','title']]

In [8]:
movies

movie_id                                      title
0            1                           Toy Story (1995)
1            2                           GoldenEye (1995)
2            3                          Four Rooms (1995)
3            4                          Get Shorty (1995)
4            5                             Copycat (1995)
...        ...                                        ...
1677      1678                          Mat' i syn (1997)
1678      1679                           B. Monkey (1998)
1679      1680                       Sliding Doors (1998)
1680      1681                        You So Crazy (1994)
1681      1682  Scream of Stone (Schrei aus Stein) (1991)

[1682 rows x 2 columns]

`#U.data contain all the ratings “We see that this file gives us information regarding the movie's title, release date, IMDb URL, and its genre(s). Since we are focused on building only collaborative filters in this chapter, we do not require any of this information, apart from the movie title and its corresponding ID:”`


In [9]:
r_cols = ['user_id','movie', 'rating','time-stamp']

rating = pd.read_csv('/Users/nandrajog/Downloads/ml-100k/u.data',sep='\t', names = r_cols,encoding = 'latin-1')

In [10]:
rating.head()

user_id  movie  rating  time-stamp
0      196    242       3   881250949
1      186    302       3   891717742
2       22    377       1   878887116
3      244     51       2   880606923
4      166    346       1   886397596

In [11]:
rating.drop(columns='time-stamp',inplace=True)

In [12]:
rating

user_id  movie  rating
0          196    242       3
1          186    302       3
2           22    377       1
3          244     51       2
4          166    346       1
...        ...    ...     ...
99995      880    476       3
99996      716    204       5
99997      276   1090       1
99998       13    225       2
99999       12    203       3

[100000 rows x 3 columns]

`Training and Testing data`

`Use this as a regression problem`

`Assume user_id be a(target variable Y) and Dataframe consist of predcitior (X)`

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
X = rating.copy()
Y=rating['user_id']

In [15]:
X.head()

user_id  movie  rating
0      196    242       3
1      186    302       3
2       22    377       1
3      244     51       2
4      166    346       1

In [16]:
Y.head()

0    196
1    186
2     22
3    244
4    166
Name: user_id, dtype: int64

In [17]:
X_train,X_test,y_train,y_test = train_test_split(X,Y,test_size = 0.25, stratify=Y,random_state =42)

#### `Evaluation`

`To evaluate regression model we use RMSE and MSE`

In [18]:
from sklearn.metrics import mean_squared_error

`Function will compute mean squared error`

In [19]:
def rmse(mse):
    rm =np.square(mse)
    return rm

In [20]:
def baseline(user_id,movie_id):
    return 3.0

In [21]:
X_test

user_id  movie  rating
33745      237    489       4
93837      535    156       2
19779      176    303       3
76325       83    756       4
10309      232    204       4
...        ...    ...     ...
82918      747    466       3
86390      758    388       3
34211      198    117       1
73488      870     28       4
1828       234    808       2

[25000 rows x 3 columns]

In [22]:
# zip(X_test['user_id'], X_test['movie'])

In [23]:
# x=zip(X_test['user_id'], X_test['movie'])

In [24]:
# for (i,j) in x:
#     print(i,j)

In [25]:
# for (user,movie) in zip(X_test['user_id'], X_test['movie']):
#     y_pred = 

In [26]:
def score(cf_model):
    
    id_pairs = zip(X_test['user_id'], X_test['movie'])
    
    y_pred = np.array([cf_model(user,movie) for (user,movie) in id_pairs])
    
    y_true = np.array(X_test['rating'])
    
    mse = mean_squared_error(y_true,y_pred)
    rm = rmse(mse)
    return rm
    

In [27]:
score(baseline).round(2)

2.42

`This is what we obtain  ` RMSE = 2.42 ` , we will try to minimize the scoe`

#### ` User-Based Collaborative Filtering `

`“User-based collaborative filters do`: They find users similar to a particular user and then recommend products that those users have liked to the first User.”



`Build a matrix using pivot tabel`

In [28]:
r_matrix = X_train.pivot_table(values='rating',index = 'user_id',columns = 'movie')

In [161]:
r_matrix.T

user_id  1    2    3    4    5    6    7    8    9    10   ...  934  935  936  \
movie                                                      ...                  
1        5.0  NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN  4.0  ...  2.0  NaN  4.0   
2        3.0  NaN  NaN  NaN  3.0  NaN  NaN  NaN  NaN  NaN  ...  4.0  NaN  NaN   
3        4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  4.0   
4        3.0  NaN  NaN  NaN  NaN  NaN  5.0  NaN  NaN  NaN  ...  NaN  NaN  NaN   
5        3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1675     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1676     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1679     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1681     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1682     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

user_id  937  938  939  940  941  942  943  
movie                                       
1        NaN  4.0  NaN  NaN  5.0  NaN  NaN  
2        NaN  NaN  NaN  NaN  NaN  NaN  5.0  
3        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...      ...  ...  ...  ...  ...  ...  ...  
1675     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1676     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1679     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1681     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1682     NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[1647 rows x 943 columns]

`If we see the above cell we can observed clearly that each  row contains user and every column represent a particular movie rating.
& we can also observe that matrix is soo sparse`

##### `First approach`

` “Let's first build one of the simplest collaborative filters possible. This simply takes in user_id and movie_id and outputs the mean rating for the movie by all the users who have rated it. No distinction is made between the users. In other words, the rating of each user is assigned equal weight.`

`It is possible that some movies are available only in the test set and not the training set (and consequentially, not in our ratings matrix). In such cases, we will just default to a rating of 3.0, like the baseline model:” `



` Mean-Rating collaborative filltering `

In [30]:
def cf_user_mean(user_id,movie_id):
    
    #Check if movie exist in the r_matix
    
    if movie_id in r_matrix:
        #If movie present in r_matrix compute the mean of all the raiting givien to the movie
        
        mean_rating = r_matrix[movie_id].mean()
    else:
        mean_rating=3.0
        
    return mean_rating

In [31]:
score(cf_user_mean)

1.0972376892043978

### `Weighted Mean`

` Here we are using `

` “coefficient will be one of the similarity metrics that we computed ”`



##### Fill NA ==0 
##### cosine similarities does not work with NAN 

In [32]:
r_matrix_dummy = r_matrix.fillna(0)

In [33]:
r_matrix_dummy.head()

movie    1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0  ...   
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
5         0.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   

movie    1669  1670  1671  1673  1674  1675  1676  1679  1681  1682  
user_id                                                              
1         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
5         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[5 rows x 1647 columns]

In [34]:
r_matrix.head()

movie    1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   2.0  ...   
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
5         NaN   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   

movie    1669  1670  1671  1673  1674  1675  1676  1679  1681  1682  
user_id                                                              
1         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
5         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 1647 columns]

In [35]:
from sklearn.metrics.pairwise import cosine_similarity

In [36]:
cosine_simi = cosine_similarity(r_matrix_dummy,r_matrix_dummy)

In [37]:
cosine_similarity_df = pd.DataFrame(cosine_simi,index=r_matrix_dummy.index,
                                   columns=r_matrix_dummy.index)

In [38]:
cosine_similarity_df

user_id       1         2         3         4         5         6         7    \
user_id                                                                         
1        1.000000  0.118076  0.029097  0.011628  0.264677  0.312419  0.308729   
2        0.118076  1.000000  0.099097  0.107680  0.034279  0.152789  0.086705   
3        0.029097  0.099097  1.000000  0.252131  0.026893  0.062539  0.039767   
4        0.011628  0.107680  0.252131  1.000000  0.000000  0.045543  0.078812   
5        0.264677  0.034279  0.026893  0.000000  1.000000  0.202843  0.299619   
...           ...       ...       ...       ...       ...       ...       ...   
939      0.084456  0.186347  0.018418  0.000000  0.066000  0.092090  0.100625   
940      0.293293  0.168034  0.096993  0.132900  0.156172  0.261859  0.233843   
941      0.056765  0.106748  0.109631  0.142798  0.115842  0.097606  0.039199   
942      0.103536  0.136796  0.092574  0.097066  0.124297  0.206104  0.224227   
943      0.326491  0.080358  0.018987  0.015176  0.267574  0.187637  0.296332   

user_id       8         9         10   ...       934       935       936  \
user_id                                ...                                 
1        0.224269  0.026017  0.286411  ...  0.308475  0.055872  0.197862   
2        0.078864  0.068940  0.092399  ...  0.086927  0.259636  0.289092   
3        0.089474  0.078162  0.037670  ...  0.040918  0.019031  0.065417   
4        0.095354  0.059498  0.053879  ...  0.024226  0.050703  0.056561   
5        0.163724  0.038474  0.153021  ...  0.262547  0.048524  0.048312   
...           ...       ...       ...  ...       ...       ...       ...   
939      0.129381  0.052699  0.033686  ...  0.054072  0.192212  0.247926   
940      0.188662  0.107486  0.197107  ...  0.216670  0.072077  0.124657   
941      0.121223  0.055766  0.085402  ...  0.065280  0.130688  0.174143   
942      0.083910  0.070065  0.118945  ...  0.186477  0.080097  0.053648   
943      0.273238  0.088281  0.162538  ...  0.203504  0.173612  0.121306   

user_id       937       938       939       940       941       942       943  
user_id                                                                        
1        0.131367  0.152449  0.084456  0.293293  0.056765  0.103536  0.326491  
2        0.318824  0.149105  0.186347  0.168034  0.106748  0.136796  0.080358  
3        0.055373  0.086503  0.018418  0.096993  0.109631  0.092574  0.018987  
4        0.107294  0.098892  0.000000  0.132900  0.142798  0.097066  0.015176  
5        0.022202  0.091910  0.066000  0.156172  0.115842  0.124297  0.267574  
...           ...       ...       ...       ...       ...       ...       ...  
939      0.229980  0.286229  1.000000  0.093007  0.151775  0.057504  0.121871  
940      0.173176  0.138989  0.093007  1.000000  0.101880  0.204524  0.185019  
941      0.131990  0.259213  0.151775  0.101880  1.000000  0.037286  0.077046  
942      0.110941  0.099172  0.057504  0.204524  0.037286  1.000000  0.097347  
943      0.044086  0.192836  0.121871  0.185019  0.077046  0.097347  1.000000  

[943 rows x 943 columns]

In [39]:
def cf_users_wmean(user_id, movie_id):
# for movie_id in r_matrix: 
#     print(movie_id)
    if movie_id in r_matrix:
        # Get the similarity score
        sim_scores = cosine_similarity_df[user_id]
    #         print(sim_score)

        #Get the user ratings for the movie

        m_rating = r_matrix[movie_id]
#         print(m_rating)
        
        #Extract the indices containing NaN in the rating series
        
        idx = m_rating[m_rating.isnull()].index
#         print(idx)
        #Drop NaN 
        
        m_rating = m_rating.dropna()
        #Drop cosine from corresponding
        sim_score = sim_scores.drop(idx)
#         print(sim_score)
        
        #Compute the final weighted mean
        
        wmean_rating = np.dot(sim_score,m_rating)/sim_score.sum()
#         print(wmean_rating)
    
    else:
        wmean_rating = 3.0
        
    return wmean_rating
        

In [40]:
score(cf_users_wmean)

1.0716415403727764

In [41]:
def score(cf_model):
    
    id_pairs = zip(X_test['user_id'], X_test['movie'])
    
    y_pred = np.array([cf_model(user,movie) for (user,movie) in id_pairs])
    
    y_true = np.array(X_test['rating'])
    
    mse = mean_squared_error(y_true,y_pred)
    rm = rmse(mse)
    return rm

In [42]:
#  id_pairs = zip(X_test['user_id'], X_test['movie'])
    
#     y_pred = np.array([cf_model(user,movie) for (user,movie) in id_pairs])
    

` Let's recommend items based on user based collaborative filltering `

In [43]:
#centered cosine
def standardize(row):
    new_row = (row - row.mean())/(row.max()-row.min())
    return new_row

In [44]:
r_matrix_std = r_matrix.fillna(0).apply(standardize)

In [45]:
r_matrix_std

movie        1         2         3         4         5         6         7     \
user_id                                                                         
1        0.714104  0.534252  0.755461  0.474443  0.552068  0.985578  0.573489   
2       -0.285896 -0.065748 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
3       -0.285896 -0.065748 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
4       -0.285896 -0.065748 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
5       -0.285896  0.534252 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
...           ...       ...       ...       ...       ...       ...       ...   
939     -0.285896 -0.065748 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
940     -0.285896 -0.065748 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
941      0.714104 -0.065748 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
942     -0.285896 -0.065748 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   
943     -0.285896  0.934252 -0.044539 -0.125557 -0.047932 -0.014422 -0.226511   

movie        8         9         10    ...     1669     1670     1671  \
user_id                                ...                              
1        0.053871  0.819088  0.546978  ... -0.00106 -0.00106 -0.00106   
2       -0.146129 -0.180912  0.346978  ... -0.00106 -0.00106 -0.00106   
3       -0.146129 -0.180912 -0.053022  ... -0.00106 -0.00106 -0.00106   
4       -0.146129 -0.180912 -0.053022  ... -0.00106 -0.00106 -0.00106   
5       -0.146129 -0.180912 -0.053022  ... -0.00106 -0.00106 -0.00106   
...           ...       ...       ...  ...      ...      ...      ...   
939     -0.146129  0.819088 -0.053022  ... -0.00106 -0.00106 -0.00106   
940     -0.146129  0.419088 -0.053022  ... -0.00106 -0.00106 -0.00106   
941     -0.146129 -0.180912 -0.053022  ... -0.00106 -0.00106 -0.00106   
942     -0.146129 -0.180912 -0.053022  ... -0.00106 -0.00106 -0.00106   
943     -0.146129  0.419088 -0.053022  ... -0.00106 -0.00106 -0.00106   

movie       1673     1674     1675     1676     1679     1681     1682  
user_id                                                                 
1       -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
2       -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
3       -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
4       -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
5       -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
...          ...      ...      ...      ...      ...      ...      ...  
939     -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
940     -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
941     -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
942     -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  
943     -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106 -0.00106  

[943 rows x 1647 columns]

In [46]:
# def recommendation(df):
#     cosine_sim = cosine_similarity(r_matrix_std,r_matrix_std)
    
#     return cosine_sim

In [47]:
# recommendation(r_matrix_std)

In [48]:
rating

user_id  movie  rating
0          196    242       3
1          186    302       3
2           22    377       1
3          244     51       2
4          166    346       1
...        ...    ...     ...
99995      880    476       3
99996      716    204       5
99997      276   1090       1
99998       13    225       2
99999       12    203       3

[100000 rows x 3 columns]

In [49]:
indices = pd.Series(rating.index,index=rating['user_id']).drop_duplicates()

In [50]:
indices

user_id
196        0
186        1
22         2
244        3
166        4
       ...  
880    99995
716    99996
276    99997
13     99998
12     99999
Length: 100000, dtype: int64

In [51]:
# indxxx = idxx['user_id']

In [52]:
# print(idxx)

In [180]:
def recommend_movies(Enter_the_value):
    id_pairs = zip(X['user_id'],X['movie'])
    cosine_similarity_df_ = pd.DataFrame(r_matrix_std,index=r_matrix_dummy.index,
                                   columns=r_matrix_dummy.index)
    
    for(user,movie) in id_pairs:
        if movie in r_matrix_std:
            sim_scores = cosine_similarity_df_[user]
            
            m_rating = r_matrix_std[movie]
            
            idx = m_rating.loc[m_rating.isnull()].index
            m_rating = m_rating.dropna()
#             print(m_rating)
            sim_score = sim_scores.drop(idx).sort_values(ascending=False)
            idx= indices[Enter_the_value]
            score = list(enumerate(sim_score[idx]))
            score = sorted(score,key= lambda x: x[1], reverse =True)
            score = score[1:11]
            movie =[]
            for i in score:
                movie.append(i[0])
            return rating.iloc[movie]
        

In [181]:
recommend_movies(22)

user_id  movie  rating
0       196    242       3
1       186    302       3
2        22    377       1
4       166    346       1
5       298    474       4
6       115    265       2
7       253    465       5
8       305    451       3
9         6     86       3
10       62    257       2

In [55]:
#Create Recommendation

In [56]:
# def creat_recommendation(df):
    
# for i in r_matrix_dummy:
#     print(r_matrix_dummy[i])

In [57]:
# for user in r_matrix:
#     sim_score = cosine_similarity(r_matrix[user])

`User Demographic`

`“The basic intuition behind these filter is that users of the same demographic tend to have similar tastes. Therefore, their effectiveness depends on the assumption that women, or teenagers, or people from the same area will share the same taste in movies.”`



In [58]:
user.head()

user_id  age sex  occupation zip_code
0        1   24   M  technician    85711
1        2   53   F       other    94043
2        3   23   M      writer    32067
3        4   24   M  technician    43537
4        5   33   F       other    15213

In [59]:
merge_df = pd.merge(X_train,user)

In [60]:
merge_df

user_id  movie  rating  age sex  occupation zip_code
0          889    684       2   24   M  technician    78704
1          889    279       2   24   M  technician    78704
2          889     29       3   24   M  technician    78704
3          889    190       3   24   M  technician    78704
4          889    232       3   24   M  technician    78704
...        ...    ...     ...  ...  ..         ...      ...
74995      631    315       4   18   F     student    38866
74996      631    301       4   18   F     student    38866
74997      631    272       4   18   F     student    38866
74998      631    294       3   18   F     student    38866
74999      631    332       3   18   F     student    38866

[75000 rows x 7 columns]

In [61]:
X_train

user_id  movie  rating
94871      889    684       2
45252      135    203       4
90833      774    674       2
1570       161     98       4
46935      506    578       3
...        ...    ...     ...
55738       10    499       4
5602        73   1149       4
90648      877     52       4
94661      608    187       4
78294      834    315       5

[75000 rows x 3 columns]

In [62]:
merge_df.head()

user_id  movie  rating  age sex  occupation zip_code
0      889    684       2   24   M  technician    78704
1      889    279       2   24   M  technician    78704
2      889     29       3   24   M  technician    78704
3      889    190       3   24   M  technician    78704
4      889    232       3   24   M  technician    78704

` Compute the mean rating of each movie by gender `

In [63]:
gender_mean = merge_df[['sex','rating','movie']].groupby(['movie','sex'])['rating'].mean()

In [64]:
pd.DataFrame(gender_mean)

rating
movie sex          
1     F    3.827586
      M    3.918919
2     F    3.230769
      M    3.228916
3     F    2.785714
...             ...
1675  M    3.000000
1676  M    2.000000
1679  M    3.000000
1681  M    3.000000
1682  M    3.000000

[3048 rows x 1 columns]

In [65]:
user

user_id  age sex     occupation zip_code
0          1   24   M     technician    85711
1          2   53   F          other    94043
2          3   23   M         writer    32067
3          4   24   M     technician    43537
4          5   33   F          other    15213
..       ...  ...  ..            ...      ...
938      939   26   F        student    33319
939      940   32   M  administrator    02215
940      941   20   M        student    97229
941      942   48   F      librarian    78209
942      943   22   M        student    77841

[943 rows x 5 columns]

In [66]:
user = user.set_index('user_id')

In [67]:
user

age sex     occupation zip_code
user_id                                 
1         24   M     technician    85711
2         53   F          other    94043
3         23   M         writer    32067
4         24   M     technician    43537
5         33   F          other    15213
...      ...  ..            ...      ...
939       26   F        student    33319
940       32   M  administrator    02215
941       20   M        student    97229
942       48   F      librarian    78209
943       22   M        student    77841

[943 rows x 4 columns]

In [68]:
r_matrix.head()

movie    1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   2.0  ...   
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
5         NaN   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   

movie    1669  1670  1671  1673  1674  1675  1676  1679  1681  1682  
user_id                                                              
1         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
5         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 1647 columns]

In [69]:
gender_mean

movie  sex
1      F      3.827586
       M      3.918919
2      F      3.230769
       M      3.228916
3      F      2.785714
                ...   
1675   M      3.000000
1676   M      2.000000
1679   M      3.000000
1681   M      3.000000
1682   M      3.000000
Name: rating, Length: 3048, dtype: float64

#### `Gender Based Collaborative filttering`

In [70]:
def cf_gender(user_id,movie_id):
    
    #Check if movie is exist in the matrix
    if movie_id in r_matrix:
        #Identify the gender of the user
        gender = user.loc[user_id]['sex']
#         print(gender)
        #compute the mean rating given by that gender of the movie
        if gender in gender_mean[movie_id]:
            gender_rating = gender_mean[movie_id][gender]
        else:
            gender_rating = 3.0
    else:
        gender_rating = 3.0

    return gender_rating
        
        

In [71]:
score(cf_gender)

1.138815096870716

`NOTE: RMSE was 1.07 in first approach and rmse 1.13 in second approach(In this approach we have added Demographic features)`